agentendpoint/agentendpoint.go

```diff
@@ -0,0 +1,221 @@
+		return nil
+	}); err != nil {
+		return fmt.Errorf("error calling ReportTaskComplete: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Client) runTask(ctx context.Context) {
+	for {
+		token, err := config.IDToken()
+		if err != nil {
+			logger.Errorf(err.Error())
+			return
+		}
+
+		start := &agentendpointpb.ReportTaskStartRequest{
+			InstanceIdToken: token,
+		}
+		startRes, err := c.raw.ReportTaskStart(ctx, start)
+		if err != nil {
+			logger.Errorf(err.Error())
+			return
+		}
+
+		task := startRes.GetTask()
+		if task == nil {
+			return
+		}
+
+		switch task.GetTaskType() {
+		case agentendpointpb.TaskType_APPLY_PATCHES:
+			if !patchEnabled {
+				logger.Infof("Recieved TaskType_APPLY_PATCHES but OSPatch is not enabled")
+				continue
+			}
+
+			if err := c.RunApplyPatches(ctx, task.GetApplyPatchesTask()); err != nil {
+				logger.Errorf(err.Error())
+			}
+		case agentendpointpb.TaskType_EXEC_STEP_TASK:
+			if err := c.RunExecStep(ctx, task.GetExecStepTask()); err != nil {
+				logger.Errorf(err.Error())
+			}
+		default:
+			logger.Errorf("Unknown task type: %v", task.GetTaskType())
+		}
+	}
+}
+
+func (c *Client) handleStream(ctx context.Context, stream agentendpointpb.AgentEndpointService_ReceiveTaskNotificationClient, noti chan struct{}) error {
+	for {
+		if _, err := stream.Recv(); err != nil {
+			// Return on any stream error, even a close, the caller will simply
+			// reconnect the stream as needed.
+			return err
+		}
+
+		// Only queue up one notifcation at a time. We should only ever
+		// have one active task being worked on and one in the queue.
+		select {
+		case noti <- struct{}{}:
+			tasker.Enqueue("TaskNotification", func() {
+				// Take this task off the notification queue so another can be
+				// queued up.
+				<-noti
+				c.runTask(ctx)
+			})
+		default:
+			// Ignore the notificaction as we already have one queued.
+		}
+	}
+}
+
+func (c *Client) watchStream(ctx context.Context, noti chan struct{}, token string) error {
+	req := &agentendpointpb.ReceiveTaskNotificationRequest{
+		AgentVersion:    config.Version(),
+		InstanceIdToken: token,
+	}
+
+	stream, err := c.raw.ReceiveTaskNotification(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	if err := c.handleStream(ctx, stream, noti); err != nil {
+		if err == io.EOF {
+			// Server closed the stream indication we should reconnect.
+			return nil
+		}
+		logger.Errorf("error recieving on stream: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Client) loadTaskFromState(ctx context.Context) error {
+	st, err := loadState(taskStateFile)
+	if err != nil {
+		return fmt.Errorf("loadState error: %v", err)
+	}
+	if st != nil && st.PatchTask != nil {
+		st.PatchTask.client = c
+		tasker.Enqueue("PatchRun", func() {
+			st.PatchTask.run(ctx)
+		})
+	}
+
+	return nil
+}
+
+// WaitForTaskNotification waits for and acts on any task notification indefinitely.
+func (c *Client) WaitForTaskNotification(ctx context.Context) error {
```

Comment:
WaitForTaskNotification is the entry point to watch for tasks, who is the caller?

Reply:
It will be in osconfig main.go, right now this is unused